app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -0,0 +1,148 @@
+                                LOG.debug("$keyword: $data")
+                            }
+                        } catch (e: Exception) {
+                            // gets caught and logged by caller
+                            continuation.resumeWithException(e)
+                        }
+                        createOnionService(service.torControlConnection, continuation, port)
+                    }
+                    // FIXME When we stop unplanned, we need to inform the ShareManager
+                    //  that we stopped, so it can clear its state up, stopping webserver, etc.
+                    TorService.STATUS_STOPPING -> LOG.debug("TorService: Stopping...")
+                    TorService.STATUS_OFF -> LOG.debug("TorService: Stopped")
+                }
+            }
+        }
+        app.registerReceiver(broadcastReceiver, IntentFilter(ACTION_STATUS))
+
+        Intent(app, OnionService::class.java).also { intent ->
+            startForegroundService(app, intent)
+            app.bindService(intent, serviceConnection, BIND_AUTO_CREATE)
+        }
+        // this method suspends here until it the continuation resumes it
+    }
+
+    fun stop() {
+        LOG.info("Stopping...")
+        // FIXME TorService crashes us when getting destroyed a second time
+        //  see: https://github.com/guardianproject/tor-android/issues/57
+        if (torService != null) app.unbindService(serviceConnection)
+        torService = null
+        // simply unbinding doesn't seem sufficient for stopping a foreground service
+        Intent(app, OnionService::class.java).also { intent ->
+            app.stopService(intent)
+        }
+        broadcastReceiver?.let { app.unregisterReceiver(it) }
+        broadcastReceiver = null
+        LOG.info("Stopped")
+    }
+
+    /**
+     * Creates a new onion service each time it is called
+     * and resumes the given [continuation] with its address.
+     */
+    private fun createOnionService(
+        controlConnection: TorControlConnection,
+        continuation: Continuation<String>,
+        port: Int,
+    ) {
+        LOG.error("Starting hidden service...")
+        val portLines = Collections.singletonMap(80, "127.0.0.1:$port")
+        val response = try {
+            controlConnection.addOnion("NEW:ED25519-V3", portLines, null)
+        } catch (e: IOException) {
+            LOG.error("Error creation onion service", e)
+            continuation.resumeWithException(e)
+            return
+        }
+        if (!response.containsKey(HS_ADDRESS)) {
+            LOG.error("Tor did not return a hidden service address")
+            continuation.resumeWithException(IOException("No HS_ADDRESS"))
+        }
+        // TODO do we need to wait for the service to be actually reachable?
+        //  The HTTP connection seems to hang a bit, but finds the service soon enough.
+        //  Maybe on slow connections, getting the service up takes more time?
```

Comment:
Perhaps we could count the number of copies of our HS descriptor that have been uploaded?

At the moment Briar's TorPlugin just logs descriptor upload events and doesn't do anything with them, but as part of the mailbox work I was considering counting them so we can tell when our hidden service has become reachable. We could do something similar here (via our own control connection).

We'd need to dive into the v3 hidden service spec to find out how many copies of the descriptor need to be uploaded before clients can reliably find it.

Reply:
A first improvement could be to just wait for the first successful upload. Getting and sending the onionshare link will also take time as will receiving and using the link for the receiver. So maybe letting this happen in parallel to the subsequent uploads will be a good compromise?

Reply:
Sounds good :+1:

Reply:
Let's do that in the next MR when I'll upgrade to the new TorService release using it in a dedicated process.